### ISSUE TITLE
generate dependency file as side effect

### ISSUE THREAD

**Issue by FranklinYu** [google/jsonnet#770]
This is a feature request.

I use GNU Make and the `jsonnet` utility to generate a JSON from my `.jsonnet` files. It would generate a file suitable to consumed by GNU Make, similar to [`gcc -MD`](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html). Note that it should do so only when output file is specified with `--output-file`/`-o` (otherwise it's complicated to decide the make rule target).

For example, if [the example mentioned on website](https://jsonnet.org/learning/tutorial.html#imports) is run with `-o output.json`, it would generate a file `output.d` with content

```make
output.json: imports.jsonnet martinis.libsonnet garnish.txt

martinis.libsonnet:

garnish.txt:
```

**Comment by sbarzowski** [comment]
Let me rephrase:

You have a bunch of Jsonnet files, which are used to generate a bunch of JSON output files. Jsonnet files of course import other files. If one of the Jsonnet files changes, you want to regenerate the affected output files to keep everything in sync. You use `make` to run Jsonnet, so you would like to have the dependencies recorded in a file that `make` can use.

Is that correct? 

I think it would be better to have a separate tool which generates the dependency graph through static  analysis. A static analysis tool which generates the dependency graph could be implemented quite easily with [go-jsonnet](https://github.com/google/go-jsonnet) I think. It should require is <50 lines of the "real code" (not counting the command line interface and such). It would be enough to walk the AST, enter any encountered imports and report the paths.

Doing so at runtime also makes sense, but I think the static approach is probably a better fit for `make`.

**Comment by FranklinYu** [comment]
> It should require is <50 lines of the "real code" (not counting the command line interface and such). It would be enough to walk the AST, enter any encountered imports and report the paths.

Yes, and the AST should be walked recursively. So it need to import the AST of the imported `.libjsonnet`, which might also need the AST of another lib, etc.

> I think it would be better to have a separate tool which generates the dependency graph through static analysis.

This works for me since I don’t have a very large Jsonnet structure. I’m not sure whether we should worry about performance (one-pass v.s. two-pass for lexer and parser).

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
I've implemented a version of this tool that generates a `sha256sum` file for caching purposes:

- https://gitlab.com/gitlab-com/gl-infra/jsonnet-tool/-/merge_requests/57

**Comment by pawelbeza** [comment]
@igorwwwwwwwwwwwwwwwwwwww That would be awesome!
